### ISSUE TITLE
[1.3] `check` on template instances creates infinite recursion

### ISSUE THREAD

**Issue by Zodiase** [meteor/meteor#6712]
You could try this in the `todos` demo app.

```JavaScript
Template.App_body.onCreated(function appBodyOnCreated() {
  // This line creates an infinite loop in Meteor 1.3.
  check(this, Match.Optional(Blaze.TemplateInstance));
});
```

In one of my packages I have to check if an optional parameter is an instance of `Blaze.TemplateInstance` and this is what I've been doing. I didn't encounter this error before Meteor 1.3.

**Comment by Zodiase** [comment]
Error in console:
![image](https://cloud.githubusercontent.com/assets/1815174/14302093/e61885c2-fb62-11e5-857c-a2e455dc78a2.png)

**Comment by abernix** [comment]
I don't have a fix, but I thought I'd put what I did find here.

This is happening only with `Match.Optional` (maybe with `Match.Maybe`, but I didn't check) and is occurring because of `EJSON`, not because of `check`.  It's happening when trying to serialize the `Blaze.TemplateInstance` with `EJSON.stringify`.

This is the line throwing the error.  If you remove the `EJSON.stringify` from the error message, it works fine...
https://github.com/meteor/meteor/blob/22e5a2611d7fd78abb4aeaddc2615bcb33adc0a9/packages/check/match.js#L169

**Comment by Zodiase** [comment]
@abernix 
I know the actual infinite recursion is caused by `EJSON.stringify`ing `Blaze.TemplateInstance`, but I believe the cause is still within `check`. Reason being: `check(this, Blaze.TemplateInstance)` works perfectly fine (except it doesn't allow `null`), which means `check` doesn't always have to call `EJSON.stringify`. So it is only logical to me that when using `Match.Optional`, after ruling out the "optional" part of the logic, it will be the exact same scenario as not having the "optional".

**Comment by abernix** [comment]
I would argue that the infinite recursion in `EJSON.stingify` should be fixed too, but it should be a separate issue.   That being said, I have found what created the problem in `check`.  It has to do with the way that `Match.Maybe` was added in 1.3 (via #6220). 

I will try to get a PR open today, but I want to write a few more tests first because there was completely unintended change to `Match.Optional` that went unnoticed (it accepts `null` now! – so identical to the new `.Maybe`!).
